### PR TITLE
Filter default opaque ports for pods and services

### LINKED ISSUE #6729: filter default list of `opaque-ports` during proxy injection
#6719 changed the proxy injector so that it adds the default list of opaque ports to pods and services. While this covers most cases, there is an edge case that it would currently miss.

If the service's spec exposes ports with a `targetPort`, then blindly copying the default list of opaque ports would not cover this case. For example, if the service maps `80` -> `3306`, then it's opaque ports should account for this; `3306` should be substituted with `80` since that is the port that is opaque.

While fixing this, it would also be worth making sure the default list for pods _and_ services only takes into account ports that are exposed by the workloads spec. If a workload does not expose any ports in the default list, then the `opaque-ports` list should not be added.

After this change, the proxy injector should properly annotate the following manifest:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: svc
  namespace: kube-public
spec:
  selector:
    app: test
  ports:
  - name: foo
    port: 9090
    targetPort: 3306
  - name: bar
    port: 5432
    targetPort: 9091
  - name: baz
    port: 9092
    targetPort: not-an-int
---
apiVersion: v1
kind: Pod
metadata:
  name: foo
  labels:
    app: test
spec:
  containers:
  - name: foo
    image: mysql
    ports:
    - containerPort: 3306
```

`service/svc` annotation: `9090,5432`
`pod/foo` annotation: `3306`

### PR DESCRIPTION
**Pull request by kleimkuhler**
#6719 changed the proxy injector so that it adds the `config.linkerd.io/opaque-ports` annotation to all pods and services if they or their namespace do not already contain the annotation. The value used is the default list of opaque ports—which is `25,443,587,3306,4444,5432,6379,9300,11211` unless otherwise specified by the user during installation.

Closes #6729

The main issue with this is that if a service exposes a service port `9090` that targets `3306`, the service _should_ have `9090` set as opaque since it targets a default opaque port, but it does not. This change ensures that services with this situation have `9090` set as opaque.

Additionally, services and pods do not need an annotation for with the entire default opaque ports list if they don't expose those ports in the first place. This change will filter out ports from the default list if the service or pod does not expose them.

### tests
I've added some unit tests that demonstrate the change in behavior and explained in the original issue #6729.

Signed-off-by: Kevin Leimkuhler <kevin@kleimkuhler.com>

### CODE REVIEW

File: pkg/inject/inject.go [hunk 1]
```diff
@@ -384,15 +384,56 @@ func (conf *ResourceConfig) CreateDefaultOpaquePortsPatch() ([]byte, error) {
...
+		// If the workload is a service, then only add the default opaque
+		// ports that are exposed as a service port, or targeted as a
+		// targetPort.
+		if conf.IsPod() {
+			filteredPorts = conf.FilterPodOpaquePorts(defaultPorts)
+		} else if conf.IsService() {
+			service := conf.workload.obj.(*corev1.Service)
+			for _, p := range service.Spec.Ports {
+				port := strconv.Itoa(int(p.Port))
+				if util.ContainsString(port, defaultPorts) {
```
Comment (adleong) [inline pkg/inject/inject.go:401]: should this case only apply if targetPort is not set?  i.e. if the port simply maps to itself as the target port.

File: pkg/inject/inject.go [hunk 2]
```diff
@@ -384,15 +384,56 @@ func (conf *ResourceConfig) CreateDefaultOpaquePortsPatch() ([]byte, error) {
...
+					filteredPorts = append(filteredPorts, port)
+				} else if util.ContainsString(strconv.Itoa(int(p.TargetPort.IntVal)), defaultPorts) {
```
Comment (adleong) [inline pkg/inject/inject.go:403]: worth remembering that we miss the case where we target a named port and a pod has a container port with that name which is a default opaque port.  But I don't think there's really anything we can do about that, users will need to manually annotate their service in that case.

Comment (kleimkuhler) [reply pkg/inject/inject.go:403]: Yep I think we should be able to address this as part of #6783. When that check runs we'll be able to match all pods and services, and should be able to issue a warning that a service is missing a port in its annotation because it's `targetPort` is a name.

File: pkg/inject/inject.go (hunk 1 above)
Comment (kleimkuhler) [reply pkg/inject/inject.go]: Ok 3931fd9 addresses this and adds some additional tests to make sure we don't set a port as opaque if it targets a port that is _not_ opaque.

File: pkg/inject/inject.go [hunk 3]
```diff
@@ -384,15 +384,65 @@ func (conf *ResourceConfig) CreateDefaultOpaquePortsPatch() ([]byte, error) {
 			return nil, err
 		}
 	} else if !conf.HasWorkloadAnnotation(k8s.ProxyOpaquePortsAnnotation) {
-		opaquePorts := conf.GetValues().Proxy.OpaquePorts
-		patch, err = conf.CreateAnnotationPatch(opaquePorts)
+		defaultPorts := strings.Split(conf.GetValues().Proxy.OpaquePorts, ",")
+		var filteredPorts []string
```
Comment (mateiidavid) [inline pkg/inject/inject.go:388]: Definitely _very_ nitpicky, but perhaps this function can be simplified a bit? It's not necessarily hard to follow, but the nesting level does make it a bit more difficult to understand.

wdyt about inverting some branches and doing away with some redundant error checking? For example on L382 -- we error check but then return the error to the caller if it is not nil anyway (presumably the caller will check the error again), we could instead just return straight away there.
```go
func (conf *ResourceConfig) CreateDefaultOpaquePortsPatch() ([]byte, error) {
  if opaquePorts, ok := conf.GetConfigAnnotation(k8s.ProxyOpaquePortsAnnotation); ok {
       // Annotation is present on namespace
      // let caller deal with the error
     return conf.CreateAnnotationPatch(opaquePorts)
  }

  var patch []byte
  var err error
  if conf.HasWorkloadAnnotation(...) {
       // annotation is already present, return (nil, nil)
      // helps us reduce the nesting
     return patch, err
  }

  // past this point everything would stay the same
```

This might make it easier to follow? Although not a blocker from my side; also, explicit error checking has its uses and having just one return at the end (where we return the patch) is definitely a good way to reason with what's happening. Also, this might get changed again soon so not sure if it's worth really spending a lot of time on it 🤷🏻

**Review by mateiidavid — APPROVED** [review]
Nice, worked well. My manifests are below if anyone is interested.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-deploy-svc
  namespace: default
spec:
  selector:
    app: nginx-deploy
  sessionAffinity: None
  type: ClusterIP
  ports:
  - protocol: TCP
    port: 80
    name: replace-me
    targetPort: 3306
  - protocol: TCP
    port: 443
    name: not-replace-me
status:
  loadBalancer: {}
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deploy
spec:
  selector:
    matchLabels:
      app: nginx-deploy
  replicas: 1
  template:
    metadata:
      labels:
        app: nginx-deploy
    spec:
      containers:
      - name: nginx
        image: nginx
        ports:
          - containerPort: 3306
          - containerPort: 443
```

To test:
```sh
# 
# pod should mark all of its container ports as opaque
# since they're defaults
# 
$ k get po nginx-deploy-6585cf757c-kppc9 -o yaml | rg 'opaque'
    config.linkerd.io/opaque-ports: 3306,443

# 
# service should have 80 and 443 as opaque
#  80 targets 3306
# 
$ k get svc nginx-deploy-svc -o yaml | rg 'opaque'
    config.linkerd.io/opaque-ports: 80,443

```

**Review by adleong — APPROVED** [review]

File: pkg/inject/inject.go (hunk 3 above)
Comment (kleimkuhler) [reply pkg/inject/inject.go]: I reduced some of the nesting and changed the name to `CreateOpaquePortsPatch` since it does not necessarily create a default list; the workload may already the annotation or it may inherit one from its namespace.
